ChrispyPeaches/FocusFriends
# Add database to FocusApp

Fixes #39 

### Problem
As a developer, I need a database in the app to store users' local data

### Solution
- Added an SQLite database context to FocusApp using entity framework
    - Added support for migrations by placing the DbContext and models for the FocusApp in a shared application used by a focusapp client project and a focusapp migrations (tools) project 
- Setup a standardization between the FocusAPI and FocusApp
    - Created base database models in `FocusCore.Models`
    - Converted the FluentAPI database model creation logic to data and relationship annotations using Entity Framework Core relationship attributes and Microsoft data annotations
        - This was to allow the base models' entity relations and constraints to be defined once in the base models instead of twice in each database context 
- Setup dependency injection on FocusApp to properly pass the database context to services and pages that require it
- As a test, setup the FocusApp db to track focus sessions tied to a hard-coded user


### Documentation
- [Architecture & Design Guidelines page](https://github.com/ChrispyPeaches/FocusFriends/wiki/Architecture-&-Design-Guidelines)
    - [Added Solution Dependency Overview ](https://github.com/ChrispyPeaches/FocusFriends/wiki/Architecture-&-Design-Guidelines#solution-dependency-overview)
    - [Added FocusApp > Project Structure & Folder structure for FocusApp.Shared](https://github.com/ChrispyPeaches/FocusFriends/wiki/Architecture-&-Design-Guidelines#project--folder-structure)
    - [Added FocusApp > Dependency Injection](https://github.com/ChrispyPeaches/FocusFriends/wiki/Architecture-&-Design-Guidelines#dependency-injection)
    - [Added bullet point to simple shell breakdown to explain dependency injection in the AppShell](https://github.com/ChrispyPeaches/FocusFriends/wiki/Architecture-&-Design-Guidelines#simple-shell-breakdown)
    - Combined Backend Design Guidelines into this page
- [Database Design](https://github.com/ChrispyPeaches/FocusFriends/wiki/Database-Design)
    - Renamed from Database Schema
    - [Added schema architecture](https://github.com/ChrispyPeaches/FocusFriends/wiki/Database-Design#schema-architecture)
    - [Added Technical Design section](https://github.com/ChrispyPeaches/FocusFriends/wiki/Database-Design#technical-design)
    - [Updated schema section](https://github.com/ChrispyPeaches/FocusFriends/wiki/Database-Design#schema) to hold overall architecture, sorted into sections for each database and base schema with the necessary info

### How did you test this?
Adding testing steps here since this is less of a feature and more of a framework

#### Test adding migration to FocusApp db
1. In the `FocusApp.Shared.Data.FocusAppContext` initialize method, add the line of code below above `Database.Migrate()`
    ```
    Database.EnsureDeleted();
    ```
2. Run the FocusApp, and after the app starts, close the app
3. Apply all migrations to the migrations database in the FocusApp.Tools project[ by following this guide](https://github.com/ChrispyPeaches/FocusFriends/wiki/Entity-Framework-Core-%E2%80%90-Code-First#applying-all-pending-migrations)
4. Add a property to one of the models (it'd be easiest to add to the Badge or Pet models) in the `FocusApp.Shared.Models` namespace.
5. [Follow this guide](https://github.com/ChrispyPeaches/FocusFriends/wiki/Entity-Framework-Core-%E2%80%90-Code-First#creating-a-migration) to add a migration to the FocusApp
6. Find a part of the FocusApp.Client (I used the TimerPage constructor) and add logic to:
    1. Add an entry to the updated model (with a value used for the new property)
    2. Save the changes to the database
    3. Query the database for that entry
8. Run the FocusApp.Client in debug with a breakpoint before the added logic, the follow through the logic to where you can view the data of the queried entry
9. Verify that entry includes the value for the added property

For ease of use, here's an example where a string property called "TestProperty" was added to the Pet model:
```
public TimerPage(IAPIClient client, ITimerService timerService, FocusAppContext context)
{
    _client = client;
    _timerService = timerService;

    context.Pets.Add(new Shared.Models.Pet()
    {
        Name = "Cat",
        Price = 100,
        TestProperty = "Test"
    });
    context.SaveChangesAsync().Wait();

    var petList = context.Pets.ToList();

...
```
####
